### ISSUE TITLE
setInputType doesn't handle no-OutputLayer situations

### ISSUE THREAD

**Issue by Joshuaalbert** [deeplearning4j/deeplearning4j#2215]
Continuing from the conclusion of issue (https://github.com/deeplearning4j/deeplearning4j/issues/22109) it is found that when you specify input type in a computation graph with setInputType, then if the layer which you specify as output is not OutputLayer type you are not certain to have the correct output shape.

E.g. if input is 3d and output layer is DenseLayer (for RL case not an RnnOutputLayer), then the output shape is 2d. 

Example (https://gist.github.com/Joshuaalbert/14c440263272b484a8dbb86b73dc9cf0)

**Comment by AlexDBlack** [comment]
So, this is actually somewhat of an edge case:
softmax output on RNN (time series) activations, where we need to apply a preprocessor after the last layer.
By definition, *pre*processors come before a layer.

I've opened an issue here, but we probably won't prioritize fixing that (it just doesn't come up that often...)
https://github.com/deeplearning4j/deeplearning4j/issues/2218

Here's a workaround: not the most elegant thing ever, but it should work ok
```java
//Previously:
//        confBuilder.addLayer("output", new DenseLayer.Builder()
//            .nIn(conf.numSumNodes).nOut(numOutput)
//            .activation("softmax").build(), "sumLayer"+(conf.numSumLayers-1));
//        confBuilder.setOutputs("output");

//Replace with:
confBuilder.addLayer("output", new DenseLayer.Builder()
	.nIn(conf.numSumNodes).nOut(numOutput)
	.activation("softmax").build(), "sumLayer"+(conf.numSumLayers-1));
confBuilder.addLayer("output2", new ActivationLayer.Builder().activation("identity").build(), new FeedForwardToRnnPreProcessor(), "output");
confBuilder.setOutputs("output2");
```

Activation layer + identity activation is essentially a "do nothing" operation.

**Comment by Joshuaalbert** [comment]
Thanks for the quick fix. It's actually fairly elegent considering it's only an edit distance of 120 characters :)

**Comment by Joshuaalbert** [comment]
Actually, there are some issues with the correct determination of the batchsize in the preprocessor.

**Comment by Joshuaalbert** [comment]
In `org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doBackward(LayerVertex.java:119)`
```
 if(layerPreProcessor != null){
            INDArray eps = pair.getSecond();
            eps = layerPreProcessor.backprop(eps,graph.batchSize());
```

and batchSize calls `inputs[0].size(0);` and fails because input is null for some reason. The only difference in the code is the addition of the activation layer

**Comment by AlexDBlack** [comment]
I did run your code with that modification, without issue. Any way I can reproduce that?

**Comment by Joshuaalbert** [comment]
This is actually during a training step which is not in the code example.

**Comment by Joshuaalbert** [comment]
I just added a `cg.backpropGradient(outputs[0]);` in evaluate which reproduces it.

**Comment by AlexDBlack** [comment]
I'm assuming you have done a forward pass before calling backpropGradient?

**Comment by Joshuaalbert** [comment]
Yes, right before

**Comment by Joshuaalbert** [comment]
Also if I replace the DenseLayer+ activationLayer with a GravesLSTM with any activation the same code works.

**Comment by Joshuaalbert** [comment]
@AlexDBlack Note that at the end of rnnTimeStep that `this.inputs = null`. Then in the backprop routine when the layer is a preprocessor it tries to call graph.inputs[0].size(0) to get the batch size before layerProProcessor.backprop is run. This fails since inputs = null. This seems a little odd, since I would expect this to be a bug discovered right away. Perhaps replacing activationlayer+FF2Rnn with a DuplicateTimeSeries vertex followed by activation.

**Comment by Joshuaalbert** [comment]
DuplicateToTimeSeries + activation layer does not work. It also is appearently a preProcessor and thus calls graph.inputs[0].size(0)? Why does rnnTimeStep do this.inputs = null at the end?

**Comment by Joshuaalbert** [comment]
Solution was to add another cg.setInputs after rnnTimeStep

**Comment by AlexDBlack** [comment]
rnnTimeStep shouldn't be used at training time - it's *only* for test time.
Are you trying to approximate truncated BPTT? If so, check how it's implemented: https://github.com/deeplearning4j/deeplearning4j/blob/97bc265f33ee1e6ef26952cd42d4f5d025d94769/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/multilayer/MultiLayerNetwork.java#L1175

**Comment by Joshuaalbert** [comment]
That's what I'm doing yes :) Thanks for the link. I'm comparing now what I do to it. Seems like the rest is correct. And now it's at running smoothly.
